### ISSUE TITLE
Editor crashes on String::num right after a vcrt snprintf call

### ISSUE THREAD

**Issue by Gnumaru** [godotengine/godot#67207]
### Godot version

master branch, commit 104af08e9d7c8308c2dd6cad89a0caee774043a0

### System information

windows 10 x64

### Issue description

While loading my project (which loads correctly on the official 4.0beta2 build) the editor crashes while still on the splash screen. I tried debugging on visual studio and the crash happens on core/string/ustring.cpp, on String::num at line 1458
![image](https://user-images.githubusercontent.com/5783414/194919639-398faf36-6cf4-41d4-8441-09815635a4bb.png)

When I manually edit my project.godot and remove all autoloads, the editor gets past the start screen, but after some instants before actually populating the tree in the filesystem view, the editor crashes.

I also noticed that the crash occurs when p_decimals is negative. It passes through line 1423 "p_decimals -= (int)floor(log10(abs_num));" but abs_num is an enormous number and it's log is also huge (around 300), and when subtracting it p_decimals still continues being a negative number after reaching snprintf() later on line 1463. 

### Steps to reproduce

I haven't found a reproducible set of steps yet. even though it happens on String::num, it maybe related to some other non obvious thing, like memory management.

**Comment by Gnumaru** [comment]
The exception is actually thrown on line 1453. By adding anything bellow, the exception marker will be on the first statement after the snprintf call.

![image](https://user-images.githubusercontent.com/5783414/194948424-66d9f9bf-4805-44d7-8987-0d8a311ce14e.png)

**Comment by Gnumaru** [comment]
Them problem seems to be using very large values formated with "%lf". When using values greater than about 1E240 and formatted with "%lf" snprintf (at least msvc implementation) throws and exception.

![image](https://user-images.githubusercontent.com/5783414/194950348-33c81f55-6b92-43ec-a867-e59ddee88163.png)

**Comment by Gnumaru** [comment]
That's it. "snprintf(buf, 256, "%lf", 1E247);" does not crash (msvc 2022), but "snprintf(buf, 256, "%lf", 1E248);" does. it must be an implementation detail (or just a plain bug) on microsoft's 2022 vcrt that the maximum expected double value is much smaller than the DBL_MAX (which is 1.7976931348623158e+308)

**Comment by akien-mga** [comment]
Isn't just that the specified size of 256 chars is too small a buffer to store up to 309 digits left of the decimal, and 6 (or whatever specified) after?

Crashing on 1E248 does sound like 249 chars + decimal + 6 = 256 chars i.e. buffer overflow.

**Comment by timothyqiu** [comment]
That `snprintf()` should write at most 256 - 1 characters, so it sounds like a vcrt bug to me.

But the alternate version using `sprintf()` would crash for it anyway :sweat_smile:

**Comment by Gnumaru** [comment]
Somehow I failed to see the obvious, indeed it is a bufer overflow problem. BUT, at the same time, it is a vcrt bug. The vcrt reference

https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-170

States that "The snprintf function truncates the output when len is greater than or equal to count, by placing a null-terminator at buffer[count-1].". So the expected behavior, acording to the online reference, should be safely truncating the output, not overflowing the buffer.

I have made a PR (#67247) to fix this.
